### PR TITLE
added optional allow empty commits

### LINKED ISSUE #7042: Allow forced empty commits and merges
Allow empty commits by setting some optional "allow-empty" property.  This should be possible both on both commits and merges.  **Keep backwards compatibility and be like Git:** By default continue to _forbid it_, only allow it if explicitly requested.

## Uses

* Any periodic commit operation: the empty commit records the fact that the process occurred and committed, and that it added no new data.
* Update mtime or other undigested fields, to create a new commit.
* Simplify automated tasks: it is often easier to commit than to commit and then handle one particular error.
* (Merges): Record that a merge was tried but had nothing to do.

## 🦕 Ancient history 🦖 

Note that this is partially the reverse of #1172.

<details>
 (I like to think of that set of fixes before we released Graveler as a group effort, please don't look to see who opened it)
</details>

  At the time we wanted to keep the Git behaviour of failing an empty commit by default.  IIRC we didn't want to add an "allow empty commits" flag, for fear that it would never be used.

Well, now we have a user :-)

## Context: Slack discussion

This [user request here](https://lakefs.slack.com/archives/C016726JLJW/p1700736838542909?thread_ts=1700640298.594399&cid=C016726JLJW) is for something potentially larger, but by allowing just this we can get the desired behaviour.  I believe that this proves it's great tooling!

### PR DESCRIPTION
**Pull request by lynnro314**
Closes #7042

### CODE REVIEW

File: pkg/graveler/graveler.go [hunk 1]
```diff
@@ -436,6 +436,7 @@ type CommitParams struct {
 	Metadata Metadata
 	// SourceMetaRange - If exists, use it directly. Fail if branch has uncommitted changes
 	SourceMetaRange *MetaRangeID
+	AllowEmpty      *bool
```
Comment (johnnyaug) [inline pkg/graveler/graveler.go:439]: Why a pointer?

File: pkg/api/controller.go [hunk 2]
```diff
@@ -2520,7 +2520,7 @@ func (c *Controller) Commit(w http.ResponseWriter, r *http.Request, body apigen.
 		metadata = body.Metadata.AdditionalProperties
 	}
 
-	newCommit, err := c.Catalog.Commit(ctx, repository, branch, body.Message, user.Committer(), metadata, body.Date, params.SourceMetarange, graveler.WithForce(swag.BoolValue(body.Force)))
+	newCommit, err := c.Catalog.Commit(ctx, repository, branch, body.Message, user.Committer(), metadata, body.Date, params.SourceMetarange, body.AllowEmpty, graveler.WithForce(swag.BoolValue(body.Force)))
```
Comment (nopcoder) [inline pkg/api/controller.go:2523]: use swag.BoolValue to pass allow empty to the underlying layer. we no longer need pointer to represent optional.

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -1110,7 +1110,7 @@ func (c *Catalog) ResetEntries(ctx context.Context, repositoryID string, branch
 	return c.Store.ResetPrefix(ctx, repository, branchID, keyPrefix, opts...)
 }
 
-func (c *Catalog) Commit(ctx context.Context, repositoryID, branch, message, committer string, metadata Metadata, date *int64, sourceMetarange *string, opts ...graveler.SetOptionsFunc) (*CommitLog, error) {
+func (c *Catalog) Commit(ctx context.Context, repositoryID, branch, message, committer string, metadata Metadata, date *int64, sourceMetarange *string, allowEmpty *bool, opts ...graveler.SetOptionsFunc) (*CommitLog, error) {
```
Comment (nopcoder) [inline pkg/catalog/catalog.go:1113]: Think we should refactor the signature - can be in a different pr and just pass commit params/options same struct layout as we pass to graveler so it will not stay a long list of arguments. When using struct the default value of each parameter will represent the default behaviour and the caller will not pass pointers to date and source metarange.

**Review by nopcoder — CHANGES_REQUESTED** [review]

**Review by nopcoder — APPROVED** [review]
LGTM! thanks for the quick fix.

**Comment by kesarwam** [comment]
@lynnro314 Can we add this to [Python Wrapper](https://pydocs-lakefs.lakefs.io/lakefs.branch.html#lakefs.branch.Branch.commit) also?
